Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -31,6 +31,11 @@ public interface INode {
     /// losing source location information, so those transformed nodes should not be returned by this property.
     /// </summary>
     IEnumerable<INode> Children { get; }
+
+    /// <summary>
+    /// These tokens are "owned" by the node
```

Comment:
Could you add more documentation? I don't know what 'owned' means. Are these all lexical tokens starting at the first non-trivia and ending at the last non-trivia token, or the subset of those tokens that is not owned by any child AST nodes?

Reply:
> For example, in the case of `var a: int, b: int`, there are two field declarations, `var a: int` and `var b: int`. Although both of them have for start token var and neither of them is a child of each other, the var will only owned by the field declaration a for formatting purposes, and the field b will only have `,` its own type.

Why would both have `var` as a start token? I'd expect `,` to be the start token for the second declaration.

> You would still have to record StartToken and EndToken for every construct, rather than inferring them from OwnedTokens.

Jep, but that's easier right? I expect a good grammar language to enable automatically adding a start and end token whenever you map something to an AST node.

> The INode hierarchy is not exhaustive, e.g. it does not contain attributes in several places, again something else to fix.

Okay, and that is required for computing OwnedTokens based on start and end tokens.

> If I were to implement an automatic ownedTokens computation based on the start and the end, then temporarily all the tokens of attributes would be owned by the surrounding constructs, which will cause some problem in formatting sections that format things differently if they encounter { and } for example

I see.

I still feel this automatic computation would be a good idea. Could answering the following questions help us decide?
- How much work would it be to switch to that?
- Would it be easier to maintain?
- Would it decrease the amount of changes in `Dafny.atg`?

And in case we don't decide to switch to that, should we create an issue to track doing that at a later time to reduce our maintenance burden?

Reply:
I just gave a go to implementing this. Took me 3 days to make it right, but it's not as bad as I thought, and I think it's worth it. Thanks for pushing on this one.